### ISSUE TITLE
custom-properties package exports properties with invalid (for CSS custom prop) names

### ISSUE THREAD

**Issue by amalitsky** [system-ui/theme-ui#1728]
This could be a feature (or documentation update) request.

**Describe the bug**
Technically this bug is about `custom-properties` package but it stems from theme property names.

**To Reproduce**
To reproduce this issue we can follow the example form the [Usage section](/system-ui/theme-ui/tree/develop/packages/custom-properties/README.md#Usage) of the package readme file.

The only thing to add is some kind of logging, i.e.:
```ts
module.exports = () => {
  const customProperties = toCustomProperties(theme, 'theme-ui') // prefix provided

  console.log(customProperties) // the only line added

  return customProperties
}
```

In the corresponding log you will see generated object properties with names like these (`theme-ui` prefix was used):
```
  '--theme-ui-size-1/2': '50%',
  '--theme-ui-size-1/3': '33.333333%',
  '--theme-ui-size-2/3': '66.666667%',
  ...
  '--theme-ui-button-3D-&:hover-transform': 'translateY(-1px)',
  ...
  '--theme-ui-style-th-': first-child-pl: 0, << different issue here: value itself is invalid
```

While such property names are perfectly fine for **JS**/**TS** objects they can't be used as custom **CSS** properties wo modification (or filtering out) cause custom CSS property names [can't contain certain special characters](https://www.w3.org/TR/CSS22/syndata.html#value-def-identifier). `&` and `/` in particular.

**Expected behavior**
To be honest I'm not sure what behavior to expect here. Some property names could be modified (by replacing special chars with `-` for example) or omitted automatically.

May be nothing needs to be done besides adding a note in the _README_ regarding recommended transformation/filtering steps required before/if the output (`customProperties`) object is used for setting custom CSS props on the HTML document.

Not sure if there is any public documentation on this package besides README and #1234. Hence it is hard to tell if dumping the output into `<style>` element is common/supported use case. However naming of the package and the fact that every property gets automatically prefixed with `--` implies that it was at least considered.

**Screenshots**

![scheenshot from chrome dev tools](https://user-images.githubusercontent.com/1457274/117558295-ed3c4180-b030-11eb-81ff-67ea22b7c152.png)

_Chromium v. 90.0.4430.93 on Ubuntu_

**Additional Context**

This should not matter but for my particular scenario (_Gatsbyjs_ website) I was using _theme-ui_ packages this way:

```ts
import { merge } from 'lodash';
import { Theme } from 'theme-ui';
// @ts-ignore-next-line
import { toTheme } from '@theme-ui/typography';
import makeCustomProperties from '@theme-ui/custom-properties';
// @ts-ignore-next-line
import { tailwind } from '@theme-ui/presets';

import { options as typographyOptions } from '../utils/typography';

const styles = {} as Record<string, Record<string, any>>;

const themeWoStyles = merge(
  tailwind,
  themeUITypography,
);

// apparently `styles` make it into custom properties export
const customProps = makeCustomProperties(themeWoStyles, 'theme-ui');

// exporting custom properties into HTML through style element
styles.root = {
  ...customProps,
};

const theme = merge(
  themeWoStyles,
  { styles },
);

// eslint-disable-next-line import/no-default-export
export default theme as Theme;
```

**Comment by amalitsky** [comment]
Folks, I can work on the fix provided some guidance regarding expected behavior.

**Comment by hasparus** [comment]
Hey @amalitsky, thanks for the issue!

I could introduce you to the repo or help you with the implementation, but quite frankly I have no opinion on how it should work.

@lachlanjc @atanasster @fcisio? What do you think?

**Comment by fcisio** [comment]
Hmmm, I never really explored this use case. I did look into it briefly, but I guess I never really needed it.

I also use Theme UI with Gatsby, and I often have very specific use cases to which I have to find theme-ui solutions.
If you want, you can join Theme UI on Discord and DM me. We could discuss your specific use case and see if the feature is truly what you need.

As for how `makeCustomProperties` should behave, I think in this case it's just a matter of controlling what you feed into the function. If you feed the entire theme everything seems to generate Custom Properties (object styles, scales, config, variants).

You could try splitting your theme into multiple files and importing all the parts in the index.
For example, you could have a `scales.js` file that only contains scales, (same for colors, ...)

---

Then the last remaining issue would be the illegal character in the custom properties.
For that, I think it's more so the user's job to handle that.

```js
sizes: {
  '1/2': '50%',  // nope
  '1-2': '50%', // yes
}
```
> We could replace illegal characters in the package, but it would be confusing for the user to have different syntax for the same values.

I definitely think there are better alternatives to achieve what you are describing. Again, feel free to DM me.

A quick example of what I'm talking about before I finish up.

> The size object really shouldn't be used with fractions as keys. It's more so meant for specific values (sidebar, icons, ...)

You can achieve something better like so:

```jsx
<Component sx={{ '--size': [1/2, 1/4], width: "calc(100% * var(--size))" }}/>
```

**Comment by amalitsky** [comment]
@fcisio, thanks for offering help, I'll ping you on Discord later on. For the sake of the thread I want to add couple of comments.

My use case is quite specific but in more generic terms it can be described as **getting one of default themes (from _presets_ package), potentially modifying it and "dumping" into DOM as CSS custom properties**. Is it a "valid" use case? By valid I mean something worth supporting "out of the box"?

If it is valid - we can proceed with discussions on what can be done to improve the library user experience for that case (and may be related) ones.

If not - we could discuss if adding more info on use cases actually supported (and not) by _custom-properties_ package to the package _README_ would do any good. As of now _custom-properties_ package looks almost like some "internal" package, not even mentioned on the website.

Thanks for the comment regarding splitting theme up into multiple files. If there is a naming convention followed by at least official _preset_ themes it might be worth filtering out `styles` for instance when "whole" theme is being passed? Cause those aren't "variables" worth exporting to CSS custom properties (as I see it).

Also I (as a user of the library) don't have control over theme properties naming coming from _presets_ package, do I? Should we consider naming such as `size['1/2']` presets module (_tailwind_ in particular) issue? [Respective code](https://github.com/system-ui/theme-ui/blob/develop/packages/preset-tailwind/src/index.ts#L420-L446).

> We could replace illegal characters in the package, but it would be confusing for the user to have different syntax for the same values

Completely agree!

**Comment by fcisio** [comment]
We discussed a little bit, and I think that we can safely close this issue.

The use case for this was to use SCSS to handle all CSS instead of Emotion and Theme UI.
It's not a use case that we should see really often.

The package function as it should, if the user filters what is fed into the custom properties generator.

> Apart from illegal characters, from some of the presets. 

@amalitsky did you have more you wanted to discuss here or we can close the issue?

**Comment by amalitsky** [comment]
Big thanks for your help and time spend with me on discord, @fcisio! Really appreciate it.

Yes, personally I'm good and can achieve what I need with filtering properties or writing my own ui-theme-to-css-custom-props export altogether. Please feel free to close the issue.

I would suggest to update package's README file to reflect that passing all the theme (i.e. tailwind from presets) into `makeCustomProperties` method might not be a good idea because of this and that (`styles`, invalid property names and some values) but I leave that decision up to you guys.

**Comment by hasparus** [comment]
I really dislike the _"silent failure"_ that's currently happening, and I think we have two ways to move forward.

1. Escape CSS Custom Properties to emit to make sure we always emit valid properties, and document the transformation in the docs.
2. Document the corner case and declare it as unsupported in the docs. Print an error to user's console if `process.env.NODE_ENV === "development"`.

**Comment by hasparus** [comment]
:rocket: Issue was released in [`v0.14.0-develop.7`](https://github.com/system-ui/theme-ui/releases/tag/v0.14.0-develop.7) :rocket:

**Comment by hasparus** [comment]
:rocket: Issue was released in [`v0.14.0`](https://github.com/system-ui/theme-ui/releases/tag/v0.14.0) :rocket:
